include/deal.II/sundials/ida.h

```diff
@@ -703,10 +738,61 @@ namespace SUNDIALS
      *       then last function will be attempted again
      * - <0: Unrecoverable error the computation will be aborted and an
      * assertion will be thrown.
+     *
+     * @warning Starting with SUNDIALS 4.1, SUNDIALS provides the possibility of
+     * specifying the tolerance for the resolution. A part from the tolerance
+     * only `rhs` is provided and `dst` needs to be returned.
      */
+#  if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
+    DEAL_II_DEPRECATED_EARLY
+#  endif
     std::function<int(const VectorType &rhs, VectorType &dst)>
       solve_jacobian_system;
 
+    /**
+     * Solve the Jacobian linear system up to a specified tolerance. This
+     * function will be called by IDA (possibly several times) after
+     * setup_jacobian() has been called at least once. IDA tries to do its best
+     * to call setup_jacobian() the minimum amount of times. If convergence can
+     * be achieved without updating the Jacobian, then IDA does not call
+     * setup_jacobian() again. If, on the contrary, internal IDA convergence
+     * tests fail, then IDA calls again setup_jacobian() with updated vectors
+     * and coefficients so that successive calls to
+     * solve_jacobian_system_up_to_tolerance() lead to better convergence in the
+     * Newton process.
+     *
+     * The jacobian $J$ should be (an approximation of) the system Jacobian
+     * \f[
+     *   J=\dfrac{\partial G}{\partial y} = \dfrac{\partial F}{\partial y} +
+     *  \alpha \dfrac{\partial F}{\partial \dot y}.
+     * \f]
+     *
+     * Arguments to the function are:
+     *
+     * @param[in] rhs The system right hand side to solve for.
+     * @param[out] dst The solution of $J^{-1} * src$.
+     * @param[out] n_iter the number of iterations required to solve the
+     * jacobian system
+     * @param[in] tolerance The tolerance with which to solve the linear system
+     *   of equations.
+     *
+     * A call to this function should store in `dst` the result of $J^{-1}$
+     * applied to `src`, i.e., `J*dst = src`. It is the users responsibility
+     * to set up proper solvers and preconditioners inside this function.
+     *
+     * This function should return:
+     * - 0: Success
+     * - >0: Recoverable error (IDAReinit will be called if this happens, and
+     *       then last function will be attempted again
+     * - <0: Unrecoverable error the computation will be aborted and an
+     * assertion will be thrown.
+     */
+    std::function<int(const VectorType &rhs,
+                      VectorType &      dst,
+                      int &             n_iter,
+                      const double      tolerance)>
+      solve_jacobian_system_up_to_tolerance;
```

Comment:
@luca-heltai I'm coming late to the game here, but would it be possible to use the same function name as for the KINSOL interface? There, I used `solve_with_jacobian`. I think it would be nice if we could use consistent names...